nomt/src/page_cache.rs

```diff
@@ -0,0 +1,251 @@
+    shared: Arc<Mutex<Shared>>,
+    store: Store,
+    /// The thread pool used for fetching pages from the store.
+    ///
+    /// Used for limiting the number of concurrent page fetches.
+    fetch_tp: ThreadPool,
+}
+
+struct Shared {
+    /// The unmodified pages loaded from the store.
+    ///
+    /// This cache must be always coherent with the underlying store. That implies that the updates
+    /// performed to the store must be reflected in the cache.
+    prestine: HashMap<PageId, Page>,
+    /// The pages that were modified, but not yet committed.
+    dirty: HashMap<PageId, Page>,
+    /// Pages that are currently being fetched from the store.
+    inflight: HashMap<PageId, InflightFetch>,
+}
+
+impl PageCache {
+    pub fn new(store: Store, o: &Options) -> Self {
+        let shared = Arc::new(Mutex::new(Shared {
+            prestine: HashMap::new(),
+            dirty: HashMap::new(),
+            inflight: HashMap::new(),
+        }));
+        let fetch_tp = ThreadPool::new(o.fetch_concurrency);
+        Self {
+            shared,
+            store,
+            fetch_tp,
+        }
+    }
+
+    /// Retrieves the page data at the given [`PageId`] asynchronously.
+    ///
+    /// If the page is in the cache, it is returned immediately. If the page is not in the cache, it
+    /// is fetched from the underlying store and returned.
+    fn retrieve(&self, page_id: PageId) -> PagePromise {
+        let mut shared = self.shared.lock().unwrap();
+
+        // We first check in the dirty pages, since that's where we would find the most recent
+        // version of the page. If the page is not present there, we check in the prestine cache.
+        if let Some(page) = shared.dirty.get(&page_id) {
+            return PagePromise::Now(page.clone());
+        }
+        if let Some(page) = shared.prestine.get(&page_id) {
+            return PagePromise::Now(page.clone());
+        }
+        // In case none of those are present, we check if there is an inflight fetch for the page.
+        if let Some(inflight) = shared.inflight.get(&page_id) {
+            return inflight.promise();
+        }
+        // Nope, then we need to fetch the page from the store.
+        let inflight = InflightFetch::new();
+        let promise = inflight.promise();
+        shared.inflight.insert(page_id, inflight);
+        let task = {
+            let store = self.store.clone();
+            let shared = self.shared.clone();
+            move || {
+                let entry = store
+                    .load_page(page_id)
+                    .map(Arc::new)
+                    .map_or(Page::Nil, Page::Exists);
+                let mut shared = shared.lock().unwrap();
+                // Unwrap: the operation was inserted above. It is scheduled for execution only
+                // once. It may removed only in the line below. Therefore, `None` is impossible.
+                let inflight = shared.inflight.remove(&page_id).unwrap();
```

Comment:
I would prefer to set a convention that `expect` is to be used rather than naked `unwrap` in all cases.